[Mobile Payments] Extend In-Person Payments availability to the UK

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #8803 
Closes: #8804 
Closes: #8805 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
* Added UK-specific card reader config:
  * Supported plugins: WCPay (min version 4.4.0)
  * Supported card readers: WisePad 3
  * Supported payment methods: CARD_PRESENT
* Handled `AmountTooSmall` error according to UK-specific requirements (min amount is GBP 0.3 vs $0.5 in the US & CA)
* Hid option to collect payment with card reader under `IPP_UK` feature flag

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
#### Test collecting payments in the UK-based store with WisePad 3 card reader. 
You can ping me or CesarVC to add you to a British store.

Basic scenarios to test:
* Create a new order w/ products, fees, and shipping and collect payment using card reader
* Refund the order total
* Collect payment via the Payments hub from the "Menu" page
* Verify correct error message is shown when trying to collect too small value (< 0.3 GBP)

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

kidinov committed:
Added UK to the list of supported countries

kidinov committed:
Changed a country code used from UK to GB

samiuelson committed:
Update card reader for the UK

samiuelson committed:
Update supported plugins for the UK

samiuelson committed:
Introduce feature flag

samiuelson committed:
Handle UK-specific `AmountTooSmall` error

samiuelson committed:
Clean up `IPP_UK` feature flag check

samiuelson committed:
Merge branch 'trunk' into add-uk-support

samiuelson committed:
Update RELEASE-NOTES.txt

samiuelson committed:
Update `CardReaderPaymentErrorMapperTest`

* Added tests for US, CA, and UK - specific cases of the `AmountTooSmall` error
* Updated IPP banner test

samiuelson commented:
I'd appreciate reviews both from @kidinov and @AnirudhBhat since this is an important feature. Thanks 🙇🏼

kidinov commented on RELEASE-NOTES.txt:
```diff
@@ -4,6 +4,7 @@
 - [**] Fix adding variable subscription products to an order [https://github.com/woocommerce/woocommerce-android/pull/8798]
 - [*] Added RTL languages support in search fields [https://github.com/woocommerce/woocommerce-android/pull/8770]
 - [*][internal] Visual improvements to the Payments Hub Screen [https://github.com/woocommerce/woocommerce-android/pull/8768]
+- [****] Extend support of in-person card reader payments to the UK [https://github.com/woocommerce/woocommerce-android/pull/8662]
```
I may be wrong, but my understanding is that 3 stars is the maximum

samiuelson replied:
Fixed with 7dd47f7

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/SelectPaymentMethodViewModel.kt:
```diff
@@ -124,6 +128,16 @@ class SelectPaymentMethodViewModel @Inject constructor(
+        cardPaymentCollectibilityChecker.isCollectable(order) &&
+            isCardReaderPaymentEnabledByFeatureFlag(siteSettings?.countryCode)
+
+    private fun isCardReaderPaymentEnabledByFeatureFlag(countryCode: String?): Boolean {
```
We do have already a centralized place where we determine if a country is supported or not - `CardReaderCountryConfigProvider::provideCountryConfigFor`

What if we will temporarily add the check there:

```
    fun provideCountryConfigFor(countryCode: String?) =
        if (countryCode?.equals("GB") == true && !FeatureFlag.IPP_UK.isEnabled()) {
            CardReaderConfigForUnsupportedCountry
        } else {
            cardReaderConfigFactory.getCardReaderConfigFor(countryCode)
        }
```
?

I think the current solution won't work for all the places, e.g. Upsell Banner, manuals, onboarding check etc

samiuelson replied:
Good point 👍🏼  Fixed with d7b47a8f8311766dfc6bd14a2acd74295ea78743.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/payment/CardReaderPaymentErrorMapper.kt:
```diff
@@ -27,9 +36,22 @@ class CardReaderPaymentErrorMapper @Inject constructor() {
     ) = when (cardPaymentStatusErrorType) {
-        AmountTooSmall -> PaymentFlowError.AmountTooSmall
+        AmountTooSmall -> {
+            if (config is CardReaderConfigForSupportedCountry) {
```
We briefly discussed that - pdfdoF-2BQ-p2#comment-3841 We may want to use the message that comes from the SDK in this case, so there won't be a new string per country. I don't know if the API returns only English messages, but
* For now we support only English-speaking countries anyway
* Even if it's English only, I think it's still mostly fine, and just mentioning a number in this message will be good enough to understand the issue for most people
* That's something we can ask from Stripe eventually

If you still think that it's not the way to go, I would still advise extending our `CardReaderConfigForSupportedCountry` with something like

```
currencyCode = "$", // not sure if needed, it may be Android API to get the code by country code
minSupportedPayment = 50,
```

And based on that construct the error string

The idea is that the only that will be needed to support another country is to add a config

samiuelson replied:
> We may want to use the message that comes from the SDK

Displaying error messages directly from API response is a great idea. However, we need to investigate if it's possible to get localized messages. I think it would be a kind of a breaking change if we switch to showing English-only errors now (e.g. we have French-speaking clients in Canada). That's why we made the decision to use local/hardcoded error messages for this error and switch to showing dynamic error messages as an improvement in the next iteration once we confirm how to get a localized error message.

> currencyCode = "$", // not sure if needed, it may be Android API to get the code by country code
minSupportedPayment = 50,

It's a nice idea to add additional props to `CardReaderConfigForSupportedCountry`. I'll follow this approach in this PR 👍🏼

samiuelson replied:
Refactored with 826ae04.

kidinov reviewed (commented):
Tests good! I left a couple of comments one of which may be blocking. Please take a look

samiuelson committed:
Improve feature flag usage

samiuelson committed:
Update RELEASE-NOTES.txt

samiuelson committed:
Make min charge amount value configurable

1. Migrated from @StringRes to PaymentFlowError in PaymentFlowError
2. Updated CardReaderConfigForSupportedCountry to include minimumAllowedChargeAmount prop
3. Updated strings
4. Updated tests

samiuelson committed:
Fix code formatting

samiuelson committed:
Remove deprecated strings

samiuelson commented:
> Tests good! I left a couple of comments one of which may be blocking. Please take a look

Thanks a lot for the review 🙇🏼 I've added the suggested improvements and refactored `CardReaderConfigForSupportedCountry` to include `minimumAllowedChargeAmount` prop.

samiuelson committed:
Merge branch 'trunk' into add-uk-support

samiuelson committed:
Merge branch 'trunk' into add-uk-support

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/payment/CardReaderPaymentErrorMapper.kt:
```diff
@@ -48,4 +65,23 @@ class CardReaderPaymentErrorMapper @Inject constructor() {
+
+            PaymentFlowError.AmountTooSmall(UiStringText(message))
+        } else {
+            logger.e(
```
I think we should crash here because we are clearly in the wrong state here. We should not start connection or payment flow in a not supported country. Or am I missing something here?

Ideally, we should make it compile time check and pass to the flows `cardReaderConfig` that is `CardReaderConfigForSupportedCountry` as it's done in `CardReaderManualsViewModel`. We can make this improvement later on

samiuelson replied:
This is a good point. I've improved this with b818540.
1. I've narrowed down arg types in `CardReaderPaymentErrorMapper` to accept only the `CardReaderConfigForSupportedCountry` type.
2. Introduced check for config type in `CardReaderPaymentViewModel` right after the call to retrieve it from `cardReaderConfigProvider`. I'm not sure if crashing the app is a good idea even though it's obviously in the wrong state. Instead, I'm emitting `PaymentDeclinedErrorType` and logging an error. 
Let me know what you think, @kidinov!

kidinov replied:
Thanks!

I'd personally crash, otherwise, we just hide the problem and if people end up in this state their payment won't work and that will take as quite some time to get this feedback and determining the source of the problem will be more complicated as well

samiuelson replied:
Makes sense. Switched to crashing with 1b8e96fe185b211e8f67b928e9c0d39303639907

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/payment/CardReaderPaymentViewModel.kt:
```diff
@@ -465,13 +467,19 @@ class CardReaderPaymentViewModel
             { retry(orderId, billingEmail, it, amountLabel) }
         } ?: { initPaymentFlow(isRetry = true) }
-        val errorType = errorMapper.mapPaymentErrorToUiError(error.type)
+        val config = cardReaderConfigFactory.getCardReaderConfigFor(getStoreCountryCode())
```
Shell we use `CardReaderCountryConfigProvider` not factory here? And probably everywhere in the main module code? Otherwise you don't take into account the feature flag and other potential logic

samiuelson replied:
Good catch. Fixed with 21053e63c6067ac2f77ada944ff012622558e0be

samiuelson replied:
Updated `GetIPPFeedbackBannerData` and `ShouldShowFeedbackBannerTest` to use `CardReaderCountryConfigProvider` as well with 43a4ae8. cc @AnirudhBhat 

Looks like there aren't any other usages of `CardReaderConfigFactory` in the main module.

kidinov reviewed (commented):
LGTM! I left a couple of comments though, please take a look

<img width="329" alt="image" src="https://user-images.githubusercontent.com/4923871/232409723-714984d1-dcca-4f0d-a270-24b4fd68790c.png">

<img width="317" alt="image" src="https://user-images.githubusercontent.com/4923871/232410507-a9225b43-37e5-48e8-bb4a-1300969a5bc0.png">

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/CardReaderCountryConfigProvider.kt:
```diff
@@ -1,10 +1,23 @@
+            cardReaderConfigFactory.getCardReaderConfigFor(countryCode)
+        }
+
+    private fun isCardReaderIPPEnabledByFeatureFlagFor(countryCode: String?): Boolean {
```
We could add a test on this

samiuelson replied:
Updated the tests with b818540b8799c1bcd0af44066ebb8f6da0ee5480

kidinov commented on libs/cardreader/src/main/java/com/woocommerce/android/cardreader/config/CardReaderConfig.kt:
```diff
@@ -14,6 +15,7 @@ sealed class CardReaderConfigForSupportedCountry(
     val supportedReaders: List<ReaderType>,
     val paymentMethodTypes: List<PaymentMethodType>,
     val supportedExtensions: List<SupportedExtension>,
+    val minimumAllowedChargeAmount: BigDecimal,
```
np: we could add what is this in the name, e.g. `minimumAllowedChargeAmountInCents`

samiuelson replied:
Updated the tests with b818540b8799c1bcd0af44066ebb8f6da0ee5480

samiuelson replied:
I think this name is fine. It represents not only cents (fractional part) but also the integer part. (E.g. can be instantiated like `BigDecimal("2.50")`).

I've decided to use `BigDecimal` to keep it consistent with the type we use for "price", and "amount total" values in other parts of the app. In our case, it could be sufficient to use `Double` instead - for the sake of calling the `SiteIndependentCurrencyFormatter::formatAmountWithCurrency` method. 

Let me know if it makes sense to you, @kidinov 🙇🏼

kidinov replied:
Yep, makes sense!

samiuelson committed:
Replace `CardReaderConfigFactory` with `CardReaderCountryConfigProvider`

samiuelson committed:
Update `CardReaderCountryConfigProviderTest`

1. Use fake `CardReaderConfigFactory` instead of mocked one
2. Add test for "GB" country code

samiuelson committed:
Fix detekt issues

samiuelson committed:
Improve config type checks in `CardReaderPaymentErrorMapper`

1. Changed card reader config type requirement to `CardReaderConfigForSupportedCountry` in `CardReaderPaymentErrorMapper`'s methods
2. Introduced check for config type in `CardReaderPaymentViewModel` right after the call to `cardReaderConfigProvider`

samiuelson committed:
Fix detekt issues

samiuelson commented:
> LGTM! I left a couple of comments though, please take a look
> 
Thanks a lot, @kidinov ! I've addressed your comments. Let me know what you think 🙇🏼

samiuelson committed:
Fix config prop type in `CardReaderPaymentErrorMapperTest`

samiuelson committed:
Crash app if it's in wrong state during card payment

samiuelson committed:
Rename prop in `CardReaderPaymentViewModelTest`

samiuelson committed:
Update `CardReaderPaymentErrorMapperTest`

Remove redundant argument

samiuelson committed:
Update `ShouldShowFeedbackBanner` to use `CardReaderCountryConfigProvider`

Using `CardReaderCountryConfigProvider` instead of `CardReaderConfigFactory` to take care of IPP_UK feature flag.

samiuelson committed:
Enable `IPP_UK` feature flag by default

The feature flag will serve for turning off card payments in the UK if necessary

kidinov commented:
@AnirudhBhat I've done with the review, please take a look

samiuelson committed:
Merge branch 'trunk' into add-uk-support

AnirudhBhat reviewed (approved):
Great work @samiuelson. LGTM! :shipit:
